### ISSUE TITLE
Incorrect validation for regular expressions - no-useless-escape

### ISSUE THREAD

**Issue by IgorNovozhilov** [eslint/eslint#9370]
```
  4:39  error  Unnecessary escape character: \[  no-useless-escape
  4:43  error  Unnecessary escape character: \(  no-useless-escape
  4:45  error  Unnecessary escape character: \)  no-useless-escape
```
cfg
```json
{
  "extends": "eslint:recommended",
  "parserOptions": {
    "ecmaVersion": 2017
  },
  "env": {
    "es6": true,
    "node": true
  },
  "rules": {
    "strict": "warn",
    "indent": ["warn", 2, { "SwitchCase": 1 }],
    "semi": ["warn", "always"],
    "no-trailing-spaces": "warn",
    "quotes": ["warn", "single", { "avoidEscape": true }],
    "no-console": "off"
  }
}
```
![2017-09-29 21-44-01](https://user-images.githubusercontent.com/3082812/31031025-1c0311cc-a560-11e7-9ae9-1d6558f4ccdf.png)

**Comment by not-an-aardvark** [comment]
Hi,

ESLint is reporting an error here because you are using an unnecessary escape character. To fix the problem, you should remove the `\` in the locations where the problem is reported. Your code will work in the same way as it did before.

**Comment by IgorNovozhilov** [comment]
Unescaped special characters that are part of regular expression syntax, which is searching in the text, in my opinion does not look obvious. Are you sure this is really the correct case, even if RegExp really works, it can cause confusion when used?

**Comment by not-an-aardvark** [comment]
The `no-useless-escape` rule is designed to report useless escape characters. Your code contains useless escape characters here, so the rule is working as designed. (`[`, `(`, and `)` do not need to be escaped in regular expressions when they are part of a character class.)

**Comment by IgorNovozhilov** [comment]
:ok:

**Comment by platinumazure** [comment]
@IgorNovozhilov As with all ESLint rules, if you disagree with a rule and prefer a different style of code for readability, you can use [disable comments](https://eslint.org/docs/user-guide/configuring#disabling-rules-with-inline-comments) to temporarily disable the rule.

**Comment by vuhrmeister** [comment]
I get this error for my E-Mail regex. When I remove the escape character then the regex doesn't have the same result. So in the following case this is a false positive error:

```javascript
/^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
```

**Comment by platinumazure** [comment]
@vuhrmeister Please open a [new issue](https://github.com/eslint/eslint/issues/new?template=ISSUE_TEMPLATE.md), filling out everything that we ask for in the template, and we can investigate. Thanks!
